### ISSUE TITLE
Running docker logs consumes a large amount a storage space if the logs are compressed

### ISSUE THREAD

**Issue by lovasoa** [moby/moby#41678]
**Description**

When running `docker logs` on a container for which the logs are compressed, docker tries to decompress all log files to disk simultaneously, requiring as much disk space as if the logs were not compressed.

I have a limited amount of disk space available on the partition where I store my docker logs. So I set the logging driver to compress the logs,  set a custom `max-size` and `max-file`, and everything is working well, until I try to actually read the logs with `docker logs`. Docker tries to decompress all the files to disk directly, and doesn't delete the decompressed files until they all have been processed. As a result, it requires as much disk space as if i had set `"compress": "false"`. 

This is a surprising behavior, and I wouldn't have expected docker to decompress *any* file to the disk, but rather to read the compressed files directly and stream the results without creating any new file.    

**Steps to reproduce the issue:**
1. Activate log compression, for instance using `"log-driver": "json-file", "log-opts": { "compress": "true", "max-size": "10m",
    "max-file": "10" }` in the docker daemon configuration.
2. Create a container that emits a lot of well-compressible output. For instance : `while true; do echo xxx; done` and let it run until all the log files are created.
3. Measure the amount of disk space used by the logs.
4. Run `docker logs`
5. While docker logs is running, monitor the amount of disk space used by the logs.

**Describe the results you received:**

The amount of disk space used by the logging folder continuously grows while `docker logs` is running.

In the end, if the partition is not large enough, `docker logs` exits with the following error: `error from daemon in stream: Error grabbing logs: error getting reference to decompressed log file: error while copying decompressed log stream to file: write /home/docker/containers/CONTAINER_ID/CONTAINER_ID-json.log.6.gz.tmp: no space left on device`

**Describe the results you expected:**

Running `docker logs` doesn't create any new file and doesn't use any additional space.  

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:23:36 2020
 OS/Arch:           linux/arm
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:45:06 2020
  OS/Arch:          linux/arm
  Experimental:     false
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.9.146-rt125w02.09.01
 Operating System: PTXdist / WAGO-PFC (containerized)
 OSType: linux
 Architecture: armv7l
 CPUs: 1
 Total Memory: 492.8MiB
 Name: PFC200V3-465405
 ID: PPMV:NY5C:GQBI:NTME:3ZK5:3UXJ:37IH:LD4F:SQ2I:6PWJ:PTVF:FCVG
 Docker Root Dir: /home/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running on a device with limited disk space available.

**Comment by cpuguy83** [comment]
This is because `docker logs` requests *all* logs.
If you specify a timestamp to fetch logs for I know we optimize this.

We can probably also do better and check number of log lines... and maybe do this lazily.
Also, the decompressed size should not exceed the max log size * number of files.

**Comment by cpuguy83** [comment]
For those interested, this is handled here: https://github.com/moby/moby/blob/b6bfff2a628fe823cfe9dc6797383769aa2f9db1/daemon/logger/loggerutils/logfile.go#L385

**Comment by lovasoa** [comment]
> This is because `docker logs` requests _all_ logs.
>  If you specify a timestamp to fetch logs for I know we optimize this.

Unless I'm mistaken, when using a timestamp to tail the logs, only the log files that are needed are considered, but they are still all decompressed to disk anyway instead of just being read line by line.


> We can probably also do better and check number of log lines... and maybe do this lazily.
I am not sure I understand. What would docker do with the number of log lines ?

>  Also, the decompressed size should not exceed the max log size * number of files.
I don't think there should be *anything* written to disk while reading the logs. And when writing to disk, the problem is that the total used space is max log size * number of files + total compressed log size.

**Comment by cpuguy83** [comment]
It does not decompress if the log file does not contain logs within the time range.
We store this metadata in a gzip header.

**Comment by lovasoa** [comment]
Looking at the code, does not look like it takes the whole time **range** into account. It takes only `since`, and ignores `until`. This makes it impossible to work around this bug by repetitively calling docker logs on successive intervals.

**Comment by thaJeztah** [comment]
@cpuguy83 How does the `local` logging driver compare in this respect? Would it be more performant?

**Comment by lovasoa** [comment]
> however if the logs can't fit in the space, and you don't want to see the compressed logs, why are they being kept?

The uncompressed logs can't fit, the compressed logs can. And of course I  want to see them.

**Comment by lovasoa** [comment]
> Time range is taken into account. When "decompressFile" is called, we pass along the read config and returns an empty file of the configured time stamp is not in the compressed file.

Maybe I'm reading the source wrong, but it looks like only `since` is taken into account and not `until`, as I was saying.

https://github.com/moby/moby/blob/b6bfff2a628fe823cfe9dc6797383769aa2f9db1/daemon/logger/loggerutils/logfile.go#L511-L515

`decompressFile` takes only `since` and not `until`

**Comment by cpuguy83** [comment]
Oh yes, sorry I missed part of your comment. It is only for `since` right now.

**Comment by cpuguy83** [comment]
The reason we decompress the logs is we need to be able to scan the logs for the `--tail` flag.

So a few potential optimizations:

- Check `until` before decompressing
- Decompress within the stream for when `--tail` is not set
- Annotate the gzip to indicate the number of log lines which are in the compressed file so it can decide if it needs to decompress more things... this probably has a lot of overhead since it would need to scan the file rotate and before compression... this may not be worth the optimization
- Lazily decompress files... perhaps on `read()`.... I think it does it on `open` right now, as I recall.

**Comment by lovasoa** [comment]
> The reason we decompress the logs is we need to be able to scan the logs for the --tail flag.

I don't understand why  the `--tail` flag forces to decompress to disk. Is there something that is preventing from reading the files (both the compressed and uncompressed ones) to a channel of log lines, and implementing since, until, and tail on that channel ?

**Comment by cpuguy83** [comment]
> Is there something that is preventing from reading the files (both the compressed and uncompressed ones) to a channel of log lines

Yes, you can't seek through a gzip file so there is no way to know what a log line is until the file is fully decompressed. Lazy decompression on `read` (and subsequently releasing/removing the decompressed file as soon as we are done with it) may help.

**Comment by lovasoa** [comment]
You **can** read a gzip line by line without decompressing the entire file.  You don't need seeking for that. I just implemented it in a separate program as a workaround for this issue.

**Comment by lovasoa** [comment]
gzip offers streaming compression and decompression. To read a line, you just decompress the stream until you find a newline character.

**Comment by cpuguy83** [comment]
For tailing log files you need to read from the end of the file.

**Comment by lovasoa** [comment]
You only need reading line by line from the beginning, keeping the last n lines you have seen in memory. When you reach the end of the file, you return the n lines you currently have in memory. This is much more efficient than decompressing the whole file to disk, then iterating over it again.

**Comment by cpuguy83** [comment]
@lovasoa If you can write a patch that does this, I'm fine with it as long as the complexity not significantly higher and does not impact CPU and memory consumption negatively.

**Comment by lovasoa** [comment]
For anyone stumbling on this issue: here is a small go program that works around the bug in docker and reads the logs without decompressing them to disk: https://gist.github.com/lovasoa/de648e02b19b764399ce85ace2b460e9
